src/Identity/IdentityServer/WebAuthnGrantValidator.cs

```diff
@@ -0,0 +1,142 @@
+        IPolicyService policyService,
+        IDataProtectorTokenFactory<SsoEmail2faSessionTokenable> tokenDataFactory,
+        IDataProtectorTokenFactory<WebAuthnLoginAssertionOptionsTokenable> assertionOptionsDataProtector,
+        IFeatureService featureService,
+        IDistributedCache distributedCache,
+        IUserDecryptionOptionsBuilder userDecryptionOptionsBuilder
+        )
+        : base(userManager, deviceRepository, deviceService, userService, eventService,
+            organizationDuoWebTokenProvider, organizationRepository, organizationUserRepository,
+            applicationCacheService, mailService, logger, currentContext, globalSettings,
+            userRepository, policyService, tokenDataFactory, featureService, ssoConfigRepository, distributedCache, userDecryptionOptionsBuilder)
+    {
+        _assertionOptionsDataProtector = assertionOptionsDataProtector;
+    }
+
+    string IExtensionGrantValidator.GrantType => "webauthn";
+
+    public async Task ValidateAsync(ExtensionGrantValidationContext context)
+    {
+        var rawToken = context.Request.Raw.Get("token");
+        var rawDeviceResponse = context.Request.Raw.Get("deviceResponse");
+        if (string.IsNullOrWhiteSpace(rawToken) || string.IsNullOrWhiteSpace(rawDeviceResponse))
+        {
+            context.Result = new GrantValidationResult(TokenRequestErrors.InvalidGrant);
+            return;
+        }
+
+        var verified = _assertionOptionsDataProtector.TryUnprotect(rawToken, out var token) &&
+            token.TokenIsValid(WebAuthnLoginAssertionOptionsScope.Authentication);
+        var deviceResponse = JsonSerializer.Deserialize<AuthenticatorAssertionRawResponse>(rawDeviceResponse);
+
+        if (!verified)
+        {
+            context.Result = new GrantValidationResult(TokenRequestErrors.InvalidRequest);
+            return;
+        }
+
+        var (user, credential) = await _userService.CompleteWebAuthLoginAssertionAsync(token.Options, deviceResponse);
+        var validatorContext = new CustomValidatorRequestContext
+        {
+            User = user,
+            KnownDevice = await KnownDeviceAsync(user, context.Request)
+        };
+
+        UserDecryptionOptionsBuilder.WithWebAuthnLoginCredential(credential);
+
+        await ValidateAsync(context, context.Request, validatorContext);
+    }
+
+    protected override Task<bool> ValidateContextAsync(ExtensionGrantValidationContext context,
+        CustomValidatorRequestContext validatorContext)
+    {
+        if (validatorContext.User == null)
+        {
+            return Task.FromResult(false);
+        }
+
+        return Task.FromResult(true);
+    }
+
+    protected override Task SetSuccessResult(ExtensionGrantValidationContext context, User user,
+        List<Claim> claims, Dictionary<string, object> customResponse)
+    {
+        context.Result = new GrantValidationResult(user.Id.ToString(), "Application",
+            identityProvider: "bitwarden",
```

Comment:
Question: should "bitwarden" be a const somewhere vs a magic string?

Reply:
I was wrong, I was the first to duplicate. I've created a new Constants class in the identity server package and move the string there